src/cpu/matmul/ref_matmul.cpp

```diff
@@ -221,41 +235,118 @@ status_t ref_matmul_t::execute_ref(const exec_ctx_t &ctx) const {
     auto sum_dt = pd()->attr()->post_ops_.get_sum_dt(dst_d.data_type());
     bool with_dropout = !pd()->attr()->dropout_.has_default_values();
 
-    // computations Note: If dst type is < 8 bits, we cannot split a
-    // byte during store or we get a race condition. To simplify
-    // logic, we limit parallelization on M and N by a factor of 2.
-    parallel_nd(batch, utils::div_up(M, 2), utils::div_up(N, 2),
-            [&](dim_t mb, dim_t m_, dim_t n_) {
-                for_(int m = 2 * m_; m < std::min<int>(2 * (m_ + 1), M); m++)
-                for (int n = 2 * n_; n < std::min<int>(2 * (n_ + 1), N); n++) {
-                    dims_t dst_dims_idx;
-                    const size_t offset = mb * M * N + m * N + n;
-                    utils::l_dims_by_l_offset(
-                            dst_dims_idx, offset, dst_d.dims(), ndims);
-
-                    float d = ker(dst_dims_idx, m, n);
-                    if (bias) d += ker_bias(dst_dims_idx);
-
-                    const auto dst_off = dst_d.off_v(dst_dims_idx);
-                    if (non_default_attrs) {
-                        if (with_dropout)
-                            d = ref_dropout(
-                                    d, dropout_mask, dst_off, *p, *seed);
-                        ref_post_ops_t::args_t args;
-                        args.dst_val
-                                = io::load_float_value(sum_dt, dst, dst_off);
-                        args.ctx = &ctx;
-                        args.l_offset = offset;
-                        args.dst_md = pd()->dst_md();
-                        ref_post_ops->execute(d, args);
+    auto scratchpad = ctx.get_scratchpad_grantor();
+    float *temp_dst = scratchpad.template get<float>(
+            memory_tracking::names::key_matmul_dst_in_acc_dt);
+
+    // Note 1: If dst type is < 8 bits, we cannot split a byte during
+    //         store or we get a race condition. To simplify logic, we limit
+    //         parallelization on M and N by a factor of 2.
+    // Note 2: dynamic quantization requires to compute a reduction
+    //         along dst groups, so we split parallelization accordingly
+    dim_t M_chunk = std::max<dim_t>(2, dst_scale_group_m);
+    dim_t N_chunk = std::max<dim_t>(2, dst_scale_group_n);
+    parallel_nd(batch, utils::div_up(M, M_chunk), utils::div_up(N, N_chunk),
```

Comment:
Let's introduce `M_chunks` and `N_chunks` and rename `M_chunk` to `M_chunk_size`, so that lambda doesn't shrink to 40 symbols, it should lead to better readability.